### PR TITLE
to_html formatter not called for float values in a mixed-type column (2)

### LINKED ISSUE #13021: to_html formatter not called for float values in a mixed-type column
#### When column x has float64 type, the custom formatter is applied to every value

```
t = pd.DataFrame([[0], [10.4], [3]], columns=['x'])
display(HTML(t.to_html(formatters=dict(x='xxx'.format))))

   x
------
0  xxx
1  xxx
2  xxx
```
#### But when column x is mixed type, the formatter isn't called for (only) floating point values

```
t = pd.DataFrame([['a'], [0], [10.4], [3]], columns=['x'])
display(HTML(t.to_html(formatters=dict(x='xxx'.format))))

   x
------
0  xxx
1  xxx
2  10.4
3  xxx
```
## INSTALLED VERSIONS

commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 14.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: None
pip: 1.5.4
setuptools: 20.7.0
Cython: None
numpy: 1.11.0
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2016.3
blosc: None
bottleneck: None
tables: 3.2.2
numexpr: 2.5.2
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: None
psycopg2: None
jinja2: None
boto: None

### PR DESCRIPTION
**Pull request by simonjayhawkins**
- [x] closes #13021
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

resubmit of #25983

### CODE REVIEW

**Comment by simonjayhawkins** [comment]
considering https://github.com/pandas-dev/pandas/pull/24134#discussion_r240038431 , this PR may need to revert some of #24134 which i think caused the regression in #25984

**Comment by simonjayhawkins** [comment]
i'm going to close this for now, since while the implementation of `ExtensionArrayFormatter` relies on formatting options AND a `formatters` parameter, it is not possible to add the shortcut for custom formatters. Changing this is outside the scope of this PR so i'll circle back to this once that's resolved.

**Comment by simonjayhawkins** [comment]
@jreback ; gone with bypassing `format_array` altogether when `formatters` are passed . it's a bit hacky, but on the bright side it closes the original issue, gets round the leading space bug #26002 and gets an additional test in play.

File: pandas/io/formats/format.py [hunk 1]
```diff
@@ -932,12 +938,17 @@ def __init__(self, values, digits=7, formatter=None, na_rep='NaN',
...
         self.leading_space = leading_space
+        self.shortcut = shortcut
 
     def get_result(self):
         fmt_values = self._format_strings()
         return _make_fixed_width(fmt_values, self.justify)
 
     def _format_strings(self):
+        # shortcut
+        if self.formatter is not None and self.shortcut:
```
Comment (jreback) [inline pandas/io/formats/format.py]: why do you need shortcut at all? isn't ``self.formatter is not None`` enough?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/format.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/io/formats/format.py]: the implementation of ExtensionArrayFormatter relies a formatter parameter and not shortcutting. so yes we need a shortcut parameter from to_string etc unless the ExtensionArrayFormatter implementation is changed.

Comment (jreback) [reply pandas/io/formats/format.py]: I would rather do that, this is too hacky

Comment (simonjayhawkins) [reply pandas/io/formats/format.py]: changing the ExtensionArrayFormatter is unlikely to be trivial.

File: pandas/io/formats/format.py [hunk 2]
```diff
@@ -938,6 +938,10 @@ def get_result(self):
         return _make_fixed_width(fmt_values, self.justify)
 
     def _format_strings(self):
+        # shortcut
```
Comment (jreback) [inline pandas/io/formats/format.py]: can you give some more explanation here

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@simonjayhawkins can you merge master, what's the status?

**Comment by simonjayhawkins** [comment]
> @simonjayhawkins can you merge master, what's the status?

changing ExtensionArrayFormatter is necessary but outside the scope of this PR.

did make a change here in this PR and ci passing. I think that is more an indication of missing tests than a solution.

will do a precursor PR instead. I'm not really happy with the implementation of ExtensionArrayFormatter but don't know why it was done this way. There presumably needs to a mechanism for 3rd party creators of EA types to display a repr. not sure if this was the reason and that is the solution though.

**Comment by simonjayhawkins** [comment]
> while the implementation of `ExtensionArrayFormatter` relies on formatting options AND a `formatters` parameter, it is not possible to add the shortcut for custom formatters.

changed formatters to be a pre-formatter instead of a shortcut so this is no longer an issue. @jorisvandenbossche

**Comment by simonjayhawkins** [comment]
latest commit also closes #26002 (again), two spaces between columns now as for other DataFrame columns

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche @TomAugspurger if you want to have a look.

**Comment by jreback** [comment]
@simonjayhawkins is this orthogonal to your formatters refactor?

**Comment by simonjayhawkins** [comment]
@jreback 

The two PRs are addressing orthogonal issues, but the implementations are not 100% orthogonal.

The other PR is a more important issue. This PR addresses a case which is probably rarely encountered by users.

This should probably be put on hold while the Extension Array formatting is discussed, since this PR does change some Extension Array formatting code.

I could probably pull out a couple of these changes into a separate PR. Changing the formatter parameter to format_array to be optional is probably not contentious and adding a test for na_rep of IntegarNA would be a good thing.

**Comment by TomAugspurger** [comment]
@simonjayhawkins there's a merge conflict.

Is this ready for 0.25? Haven't had time to look at the change.

**Comment by simonjayhawkins** [comment]
> Is this ready for 0.25?

i don't believe it is wise to merge this at this time, so closing for now and will revisit at a later date.

**Pull request closed without merging**